module/VuFind/src/VuFind/View/Helper/Root/CleanHtmlFactory.php

```diff
@@ -0,0 +1,147 @@
+     * @param ContainerInterface $container     Service manager
+     * @param string             $requestedName Service being created
+     * @param null|array         $options       Extra options (optional)
+     *
+     * @return object
+     *
+     * @throws ServiceNotFoundException if unable to resolve the service.
+     * @throws ServiceNotCreatedException if an exception is raised when
+     * creating a service.
+     * @throws ContainerException if any other error occurs
+     */
+    public function __invoke(
+        ContainerInterface $container,
+        $requestedName,
+        array $options = null
+    ) {
+        if (!empty($options)) {
+            throw new \Exception('Unexpected options sent to factory.');
+        }
+
+        $this->container = $container;
+        return new $requestedName(Closure::fromCallable([$this, 'createPurifier']));
+    }
+
+    /**
+     * Create a purifier instance.
+     *
+     * N.B. This is a relatively slow method.
+     *
+     * @param array $options Additional options. Currently supported:
+     *   targetBlank  true/false  Whether to add target="_blank" to external links
+     *
+     * @return HTMLPurifier
+     */
+    protected function createPurifier(array $options): HTMLPurifier
+    {
+        $config = \HTMLPurifier_Config::createDefault();
+        // Set cache path to the object cache
+        $cacheDir
+            = $this->container->get(\VuFind\Cache\Manager::class)->getCache('object')->getOptions()->getCacheDir();
+        if ($cacheDir) {
+            $config->set('Cache.SerializerPath', $cacheDir);
+        }
+        if ($options['targetBlank'] ?? false) {
+            $config->set('HTML.Nofollow', 1);
+            $config->set('HTML.TargetBlank', 1);
+        }
+
+        // Setting the following option makes purifier’s DOMLex pass the
+        // LIBXML_PARSEHUGE option to DOMDocument::loadHtml method. This in turn
+        // ensures that PHP calls htmlCtxtUseOptions (see
+        // github.com/php/php-src/blob/PHP-8.1.14/ext/dom/document.c#L1870),
+        // which ensures that the libxml2 options (namely keepBlanks) are set up
+        // properly, and whitespace nodes are preserved. This should not be an
+        // issue from libxml2 version 2.9.5, but during testing the issue was
+        // still intermittently present. Regardless of that, CentOS 7.x have an
+        // older libxml2 that exhibits the issue.
+        $config->set('Core.AllowParseManyTags', true);
+
+        $this->setAdditionalConfiguration($config);
+        return new \HTMLPurifier($config);
+    }
+
+    /**
+     * Sets additional configuration
+     *
+     * @param HTMLPurifier_Config $config Configuration
+     *
+     * @return void
+     */
+    protected function setAdditionalConfiguration(HTMLPurifier_Config $config)
+    {
+        // Add support for deprecated tags:
+        $definition = $config->getHTMLDefinition(true);
+        $definition->addElement(
+            'details',
```

Comment:
I was surprised that `<details>` wasn't supported already and I see that the owner ezyang says it [never](https://github.com/ezyang/htmlpurifier/issues/160#issuecomment-735522214) will be, but that people might use this [extension](https://github.com/xemlock/htmlpurifier-html5).  Given that purifier's intent is security I'm a little wary of using a library that is semi-abandoned.  But maybe it's still the best option.

Reply:
I'm quite confused about the status of this project. It seems reasonably active (release within the last year, commits within the past few months) but if it really doesn't support HTML5 properly, that seems like a pretty major concern....

Reply:
Since this is all isolated to a single view helper, the overall long-term maintenance risk is pretty low... but just wanted to be sure we weren't betting on the wrong horse from the very start! :-)

Reply:
Fair enough! For whatever it's worth, it's currently at  221 630 072 installs in packagist, so at least we won't be alone. :)